nni/common/serializer.py

```diff
@@ -0,0 +1,369 @@
+import base64
+import functools
+import inspect
+from typing import Any, Union, Dict, Optional, List, TypeVar
+
+import json_tricks  # use json_tricks as serializer backend
+import cloudpickle  # use cloudpickle as backend for unserializable types and instances
+
+
+__all__ = ['trace', 'dump', 'load', 'SerializableObject']
+
+
+T = TypeVar('T')
+
+
+class SerializableObject:
+    """
+    Serializable object is a wrapper of existing python objects, that supports dump and load easily.
+    Stores a symbol ``s`` and a dict of arguments ``args``, and the object can be restored with ``s(**args)``.
+    """
+
+    def __init__(self, symbol: T, args: List[Any], kwargs: Dict[str, Any],
+                 _self_contained: bool = False):
+        # use dict to avoid conflicts with user's getattr and setattr
+        self.__dict__['_nni_symbol'] = symbol
+        self.__dict__['_nni_args'] = args
+        self.__dict__['_nni_kwargs'] = kwargs
+
+        self.__dict__['_nni_self_contained'] = _self_contained
```

Comment:
what `_self_contained` mainly used for?

Reply:
Serializable class instances are self-contained, meaning that we don't have to use `.get()` to retrieve the original instance. It behaves naturally like an original one.